File: Back End/C++/Chapter 7 - Inheritance - Extending Classes/2K230203/inheritance_singleInheritance.cpp
// Inheritance

// Single Inheritance

#include<iostream>

class B
{
    private:
        int a;          // private; not inheritable
    public:
        int b;          // public; ready for inheritance
        void get_ab();
        int get_a(void);
        void show_a(void);
};

class D : public B      // public derivation
{
    private:
        int c;
    public:
        void mul(void);
        void display(void);
};

//----------------------------------------------------------------------

void B :: get_ab(void)
{
    a = 5;
    b =  10;
}

int B :: get_a()
{
    return a;
}

void B :: show_a()
{
    std::cout << "a = " << a << "\n";
}

void D :: mul()
{
    c = b * get_a();
}

void D :: display()
{
    std::cout << "a = " << get_a() << "\n";
    std::cout << "b = " << b << "\n";
    std::cout << "c = " << c << "\n\n";
}

//----------------------------------------------------------------------

int main()
{
    D d;

    d.get_ab();
    d.mul();
    d.show_a();
    d.display();

    d.b = 20;
    d.mul();
    d.display();

    return 0;
}

File: Back End/Python/Chapter 09 - Object-oriented programming (OOPs)/Concept 03 - Inheritance/2K230319/inheritance_singleInheritance.py
# Object-Oriented programming
#
# Inheritance:
#     1. Single Inheritance

class base_class:
    def __init__(self):
        self.a = 0
        self.b = 0

    # def fun1(self, num1, num2):
    #     self.a = num1
    #     self.b = num2

    def get_ab(self, num1, num2):
        self.a = num1
        self.b = num2
        print(end="\n")

    def get_a(self):
        return self.a

    def show_b(self):
        print("b = ", self.b, end="\n")


class derived_class(base_class):
    def __init__(self):
        base_class.__init__(self)
        self.c = 0

    def multiplication(self):
        print("Inside the class:", end="\n")
        print("a = ", base_class.get_a(self), end="\n")
        base_class.show_b(self)

        self.c = self.b * base_class.get_a(self)
        return self.c

    def display(self):
        print("\nDisplaying all the values: ", end="\n")
        # print("a = ", base_class.get_a(self), end="\n")
        # print("b = ", self.b, end="\n")
        print("c = b * a ", "|", self.c, "=", self.b, "*", self.a, end="\n")


a = int(input("Enter num1 value: "))
b = int(input("Enter num2 value: "))
obj1 = derived_class()

obj1.get_ab(a, b)
print("Outside the class: a = ", obj1.get_a(), end="\n")
print("Outside the class: ", end=" ")
obj1.show_b()

obj1.multiplication()

obj1.display()

print("\nUpdating the value of b...", end="\n")
# obj1.num2 = 30
obj1.b = int(input("Enter new value for b: "))
print("Displaying value of b: ", end=" ")
obj1.show_b()

obj1.multiplication()
obj1.display()
